### ISSUE TITLE
Add NEAT baseline (neuroevolution) alongside tabular Q-learning

### ISSUE THREAD

**Issue by ogyrec-o** [ogyrec-o/gridworld-learning-lab#1]
### Summary
Add a **NEAT (NeuroEvolution of Augmenting Topologies)** baseline to `gridworld-learning-lab` as an alternative learning algorithm to the current tabular Q-learning. The goal is to keep the project **minimal, reproducible, and CLI-friendly**, while enabling side-by-side comparisons of learning curves and demo rollouts.

### Motivation / Why
This repo is intended as a small research-style sandbox focused on evaluation:
- NEAT provides a **model-free** evolutionary baseline that does not require tabular state enumerations.
- It enables comparisons between **value-based** learning (Q-learning) and **neuroevolution** under the same environment + metrics.
- It strengthens the "iterate / evaluate / improve" story: reward shaping, difficulty changes, and algorithm comparisons.

### Goals
- Implement a NEAT training mode that can solve the same gridworld task.
- Preserve **determinism** (fixed seeds, saved config, reproducible metrics).
- Reuse the existing logging pipeline (`metrics.csv`) and GIF renderer (`demo`).

### Non-goals (for this first iteration)
- No parallel/distributed training.
- No advanced novelty search.
- No obstacles-by-default changes (obstacles remain optional).
- No deep integration with external ML frameworks.

---

### New command
Add a new subcommand:
- `gridworld-learning-lab neat-train ...`

Suggested interface:
```text
gridworld-learning-lab neat-train \
  --generations 300 \
  --population 150 \
  --max-steps 96 \
  --seed 1 \
  --out runs/neat_latest
```

### Demo compatibility
After training, produce an artifact compatible with `demo`:
- `runs/neat_latest/policy.bin` (or `neat_policy.bin`)
- `runs/neat_latest/config.json` (reusing RunConfig + neat config)

Then:
```text
gridworld-learning-lab demo \
  --policy runs/neat_latest/neat_policy.bin \
  --seed 2 \
  --gif assets/demo_neat.gif
```

(Exact flag naming is flexible, but keep it consistent and simple.)

---

### 1) Policy representation
Define a `Policy` trait used by `demo`/rollouts:
- `fn act(&self, state: u32, env: &Gridworld) -> Action`

Keep Q-table policy as one implementation; add NEAT policy as another.

### 2) NEAT genotype + phenotype
Minimal NEAT components:
- Node genes: input / hidden / output
- Connection genes: (in_node, out_node, weight, enabled, innovation_id)
- Mutation: weight perturbation, add connection, add node (split connection)
- Crossover: matching/disjoint/excess genes by innovation number
- Compatibility distance + speciation
- Fitness evaluation: average return or success-based fitness
- Elitism: keep best genomes per species

### 3) Inputs / outputs (keep minimal)
Inputs (suggested, normalized to [0..1]):
- agent_x, agent_y
- goal_zone_start_x (or distance_to_goal = (goal_x0 - x))
- optional bias 1.0

Outputs:
- 5 action logits/values mapped to {Up,Down,Left,Right,Stay} via argmax.

### 4) Fitness function
Use a deterministic evaluation:
- Evaluate each genome for `k` episodes (e.g., 3–5) with fixed seeds derived from base seed.
- Fitness suggestion:
  - success bonus (e.g., +1.0 per success)
  - minus step penalty (normalize by max_steps)
  - optional shaping disabled by default for fairness

### 5) Metrics
Write a new CSV (or reuse schema with generation as episode):
- Option A: keep same columns (treat each genome-eval as an "episode") — may get huge
- Option B (recommended): create `metrics_neat.csv` with:
  - generation, best_fitness, mean_fitness, best_success_rate, best_steps

Also keep an export compatible with the existing `metrics.csv` if easy.

### 6) Artifacts
Save:
- best genome/policy as bincode
- neat config JSON
- run config JSON
- optional: species stats (counts) for debugging

### 7) Tests
Add unit tests for core invariants:
- innovation ids monotonic / stable determinism with seed
- mutation functions do not panic and keep graph valid
- forward pass deterministic and finite (no NaN)
- a tiny known topology produces expected action selection

---

## Acceptance criteria
- `cargo clippy -- -D warnings` and `cargo test` remain clean.
- `cargo run --release -- neat-train ...` produces:
  - metrics file(s)
  - a saved policy file
  - config JSON
- `cargo run --release -- demo --policy <neat_policy> ...` renders a GIF successfully.
- Deterministic behavior: rerunning with same seeds yields identical metrics (within float tolerance) and same GIF frames (byte-identical if feasible, or visually identical).

---

## Notes / Dependency policy
Implement a small in-repo NEAT.
